File: backend/taxatlas/models/enums.py
"""Shared enumerations. String enums so they serialize cleanly and survive DB round-trips."""

from __future__ import annotations

from enum import StrEnum


class JurisdictionLevel(StrEnum):
    SUPRANATIONAL = "supranational"  # e.g. EU, WTO
    COUNTRY = "country"
    STATE = "state"  # US states, Australian states, Indian states
    PROVINCE = "province"  # Canadian provinces, Chinese provinces
    TERRITORY = "territory"
    REGION = "region"  # e.g. Spanish autonomous communities, Swiss cantons
    CITY = "city"


class TaxType(StrEnum):
    VAT = "vat"
    GST = "gst"
    SALES_USE = "sales_use"
    CORPORATE_INCOME = "corporate_income"
    PERSONAL_INCOME = "personal_income"
    WITHHOLDING = "withholding"
    CAPITAL_GAINS = "capital_gains"
    DIGITAL_SERVICES = "digital_services"
    CUSTOMS_TARIFF = "customs_tariff"
    EXCISE = "excise"
    PAYROLL_SOCIAL = "payroll_social"
    PROPERTY = "property"
    TRANSFER_PRICING = "transfer_pricing"
    PILLAR_TWO = "pillar_two"  # OECD global minimum tax
    OTHER = "other"


INDIRECT_TAX_TYPES = {TaxType.VAT, TaxType.GST, TaxType.SALES_USE, TaxType.EXCISE, TaxType.CUSTOMS_TARIFF}
INCOME_TAX_TYPES = {
    TaxType.CORPORATE_INCOME,
    TaxType.PERSONAL_INCOME,
    TaxType.WITHHOLDING,
    TaxType.CAPITAL_GAINS,
    TaxType.PILLAR_TWO,
}


class RateKind(StrEnum):
    STANDARD = "standard"
    REDUCED = "reduced"
    SUPER_REDUCED = "super_reduced"
    ZERO = "zero"
    HEADLINE = "headline"  # CIT headline / combined
    FEDERAL = "federal"
    STATE_AVERAGE = "state_average"
    TOP_MARGINAL = "top_marginal"
    DIVIDENDS = "dividends"
    INTEREST = "interest"
    ROYALTIES = "royalties"
    SERVICES = "services"
    REGISTRATION_THRESHOLD = "registration_threshold"
    ECONOMIC_NEXUS_THRESHOLD = "economic_nexus_threshold"
    MINIMUM = "minimum"
    TRADE_TAX = "trade_tax"  # municipal/trade-tax burden (DE Gewerbesteuer average per Land)
    REGIONAL = "regional"  # regional surcharge/levy layered on a national tax (IT IRAP, addizionale; JP enterprise tax)
    STAMP_DUTY = "stamp_duty"  # transfer duty on property conveyance (DE GrESt, ES ITP, IN stamp duty)
    OTHER = "other"


class Confidence(StrEnum):
    VERIFIED = "verified"  # checked against primary source on as_of date
    REPORTED = "reported"  # from secondary/aggregator source
    ESTIMATED = "estimated"  # derived / approximate


class RegulationStatus(StrEnum):
    PROPOSED = "proposed"
    CONSULTATION = "consultation"
    ENACTED = "enacted"
    EFFECTIVE = "effective"
    AMENDED = "amended"
    REPEALED = "repealed"
    GUIDANCE = "guidance"
    UNKNOWN = "unknown"


class DocType(StrEnum):
    STATUTE = "statute"
    REGULATION = "regulation"
    RULING = "ruling"
    GUIDANCE = "guidance"
    DIRECTIVE = "directive"
    TREATY = "treaty"
    NEWS = "news"
    CONSULTATION = "consultation"
    OTHER = "other"


class Significance(StrEnum):
    LANDMARK = "landmark"
    SIGNIFICANT = "significant"
    ROUTINE = "routine"


class Outcome(StrEnum):
    TAXPAYER = "taxpayer"
    GOVERNMENT = "government"
    MIXED = "mixed"
    PENDING = "pending"
    REMANDED = "remanded"


class TariffMeasure(StrEnum):
    MFN = "mfn"
    PREFERENTIAL = "preferential"
    ANTIDUMPING = "antidumping"
    COUNTERVAILING = "countervailing"
    SAFEGUARD = "safeguard"
    SECTION_232 = "section_232"
    SECTION_301 = "section_301"
    IEEPA = "ieepa"
    RETALIATORY = "retaliatory"
    CBAM = "cbam"
    QUOTA = "quota"
    EXPORT_CONTROL = "export_control"
    OTHER = "other"


class MeasureStatus(StrEnum):
    PROPOSED = "proposed"
    IN_FORCE = "in_force"
    SUSPENDED = "suspended"
    EXPIRED = "expired"
    REVOKED = "revoked"
    UNDER_REVIEW = "under_review"


class SourceCategory(StrEnum):
    REGULATION = "regulation"
    COURT = "court"
    TARIFF = "tariff"
    RATES = "rates"
    NEWS = "news"


class AdapterType(StrEnum):
    RSS = "rss"
    HTML = "html"
    JSON = "json"
    FIXTURE = "fixture"
    BROWSER = "browser"  # headless Chromium renders the page, then the html selectors apply (adapters/browser.py)
    RATES_TABLE = "rates_table"  # published rate table -> observed (code, tax_type, rate_kind, value); proposals only
    NEWS = "news"  # keyword search over aggregated press via GDELT / Bing / Google News, first provider that answers (adapters/news.py)


class CrawlStatus(StrEnum):
    RUNNING = "running"
    SUCCESS = "success"
    FAILED = "failed"
    SKIPPED = "skipped"
    UNCHANGED = "unchanged"  # HTTP 304


class EntityType(StrEnum):
    RATE = "rate"
    REGULATION = "regulation"
    COURT_DECISION = "court_decision"
    TARIFF = "tariff"
    JURISDICTION = "jurisdiction"


class ChangeType(StrEnum):
    CREATED = "created"
    UPDATED = "updated"
    RATE_CHANGED = "rate_changed"
    STATUS_CHANGED = "status_changed"
    REMOVED = "removed"


class UserRole(StrEnum):
    ADMIN = "admin"
    ANALYST = "analyst"
    VIEWER = "viewer"

File: taxatlas-ui/lib/enums.ts
/* Display labels for enum values (mirrors backend/app/models/enums.py). */

export const TAX_TYPE_LABEL: Record<string, string> = {
  vat: "VAT",
  gst: "GST",
  sales_use: "Sales & use",
  corporate_income: "Corporate income",
  personal_income: "Personal income",
  withholding: "Withholding",
  capital_gains: "Capital gains",
  digital_services: "Digital services",
  customs_tariff: "Customs / tariff",
  excise: "Excise",
  payroll_social: "Payroll / social",
  property: "Property",
  transfer_pricing: "Transfer pricing",
  pillar_two: "Pillar Two",
  other: "Other",
};
export const TAX_TYPES = Object.keys(TAX_TYPE_LABEL);

export const RATE_KIND_LABEL: Record<string, string> = {
  standard: "Standard",
  reduced: "Reduced",
  super_reduced: "Super-reduced",
  zero: "Zero",
  headline: "Headline",
  federal: "Federal",
  state_average: "State average",
  top_marginal: "Top marginal",
  dividends: "Dividends",
  interest: "Interest",
  royalties: "Royalties",
  services: "Services",
  registration_threshold: "Registration threshold",
  economic_nexus_threshold: "Economic nexus threshold",
  minimum: "Minimum",
  trade_tax: "Trade tax",
  regional: "Regional",
  stamp_duty: "Stamp / transfer duty",
  other: "Other",
};

export const REG_STATUSES = ["proposed", "consultation", "enacted", "effective", "amended", "repealed", "guidance", "unknown"];
export const DOC_TYPES = ["statute", "regulation", "ruling", "guidance", "directive", "treaty", "news", "consultation", "other"];
export const SIGNIFICANCE = ["landmark", "significant", "routine"];
export const OUTCOMES = ["taxpayer", "government", "mixed", "pending", "remanded"];
export const TARIFF_MEASURES = [
  "mfn", "preferential", "antidumping", "countervailing", "safeguard", "section_232", "section_301", "ieepa",
  "retaliatory", "cbam", "quota", "export_control", "other",
];
export const TARIFF_MEASURE_LABEL: Record<string, string> = {
  mfn: "MFN",
  preferential: "Preferential",
  antidumping: "Anti-dumping",
  countervailing: "Countervailing",
  safeguard: "Safeguard",
  section_232: "Section 232",
  section_301: "Section 301",
  ieepa: "IEEPA",
  retaliatory: "Retaliatory",
  cbam: "CBAM",
  quota: "Quota",
  export_control: "Export control",
  other: "Other",
};
export const MEASURE_STATUSES = ["proposed", "in_force", "suspended", "expired", "revoked", "under_review"];
export const ENTITY_TYPES = ["rate", "regulation", "court_decision", "tariff", "jurisdiction"];
export const CHANGE_TYPES = ["created", "updated", "rate_changed", "status_changed", "removed"];
export const LEVELS = ["supranational", "country", "state", "province", "territory", "region", "city"];
/* Regions as seeded (see /stats/overview by_region). */
export const REGIONS = ["Africa", "Asia-Pacific", "Caribbean", "Central America", "Europe", "Middle East", "North America", "Oceania", "South America"];
export const CONFIDENCE = ["verified", "reported", "estimated"];
export const CRAWL_STATUSES = ["running", "success", "failed", "skipped", "unchanged"];
export const SOURCE_CATEGORIES = ["regulation", "court", "tariff", "rates", "news"];
export const ADAPTERS = ["rss", "html", "json", "fixture"];

export const LEVEL_LABEL: Record<string, string> = {
  supranational: "Supranational",
  country: "Country",
  state: "State",
  province: "Province",
  territory: "Territory",
  region: "Region",
  city: "City",
};
export const ENTITY_LABEL: Record<string, string> = {
  rate: "Rate",
  regulation: "Regulation",
  court_decision: "Court decision",
  tariff: "Tariff",
  jurisdiction: "Jurisdiction",
};
export const CHANGE_TYPE_LABEL: Record<string, string> = {
  created: "Created",
  updated: "Updated",
  rate_changed: "Rate changed",
  status_changed: "Status changed",
  removed: "Removed",
};

/* ---------------------------------------------------------------- marker tone maps (components.md §5)
   Status markers carry a shape cue plus colour; the word is always rendered next to the dot. */
export type StatusTone = "positive" | "info" | "pending" | "halt" | "negative" | "neutral";
export const STATUS_TONE: Record<string, StatusTone> = {
  // regulation status
  proposed: "pending",
  consultation: "pending",
  enacted: "info",
  effective: "positive",
  amended: "info",
  repealed: "halt",
  guidance: "neutral",
  unknown: "neutral",
  // tariff measure status
  in_force: "positive",
  suspended: "halt",
  expired: "halt",
  revoked: "negative",
  under_review: "pending",
  // crawl / run status
  running: "pending",
  success: "positive",
  failed: "negative",
  skipped: "neutral",
  unchanged: "neutral",
  disabled: "neutral",
  enabled: "positive",
  active: "positive",
  // court outcome
  taxpayer: "positive",
  government: "info",
  mixed: "info",
  pending: "pending",
  remanded: "pending",
  // misc
  verified: "positive",
  updated: "info",
  routine: "neutral",
};
export function statusTone(v: string | null | undefined): StatusTone {
  return (v && STATUS_TONE[v]) || "neutral";
}

/** Change-type glyphs: mono, one character wide. */
export const CHANGE_GLYPH: Record<string, string> = {
  created: "+",
  updated: "~",
  rate_changed: "Δ",
  status_changed: "→",
  removed: "−",
};
/** Entity glyphs for the feed meta line (mono letters, no icons). */
export const ENTITY_GLYPH: Record<string, string> = {
  rate: "%",
  regulation: "R",
  court_decision: "C",
  tariff: "T",
  jurisdiction: "J",
};

export function label(map: Record<string, string>, v: string | null | undefined): string {
  if (!v) return "—";
  return map[v] ?? v.replace(/_/g, " ").replace(/\b\w/g, (c) => c.toUpperCase());
}

export function titleCaseOptions(values: string[]): Array<{ value: string; label: string }> {
  return values.map((v) => ({ value: v, label: v.replace(/_/g, " ").replace(/\b\w/g, (c) => c.toUpperCase()) }));
}
